linux/hid.c

```diff
@@ -818,6 +823,82 @@ HID_API_EXPORT const char* HID_API_CALL hid_version_str(void)
 	return HID_API_VERSION_STR;
 }
 
+static struct hid_hotplug_context {
+	/* UDEV context that handles the monitor */
+	struct udev* udev_ctx;
+
+	/* UDEV monitor that receives events */
+	struct udev_monitor* mon;
+
+	/* File descriptor for the UDEV monitor that allows to check for new events with select() */
+	int monitor_fd;
+
+	/* Thread for the UDEV monitor */
+	pthread_t thread;
+
+	pthread_mutex_t mutex;
+	
+	int mutex_ready;
+
+	/* HIDAPI unique callback handle counter */
+	hid_hotplug_callback_handle next_handle;
+
+	/* Linked list of the hotplug callbacks */
+	struct hid_hotplug_callback *hotplug_cbs;
+
+	/* Linked list of the device infos (mandatory when the device is disconnected) */
+	struct hid_device_info *devs;
+} hid_hotplug_context = {
+	.udev_ctx = NULL,
+	.monitor_fd = -1,
+	.next_handle = FIRST_HOTPLUG_CALLBACK_HANDLE,
+	.mutex_ready = 0,
+	.hotplug_cbs = NULL,
+	.devs = NULL
+};
+
+static void hid_internal_hotplug_cleanup()
+{
+	if (hid_hotplug_context.hotplug_cbs != NULL) {
+		return;
+	}
+
+	/* Cleanup connected device list */
+	hid_free_enumeration(hid_hotplug_context.devs);
+	hid_hotplug_context.devs = NULL;
+	/* Disarm the udev monitor */
+	udev_monitor_unref(hid_hotplug_context.mon);
+	udev_unref(hid_hotplug_context.udev_ctx);
+}
+
+static void hid_internal_hotplug_init()
+{
+	if (!hid_hotplug_context.mutex_ready) {
+		pthread_mutex_init(&hid_hotplug_context.mutex, NULL);
+		hid_hotplug_context.mutex_ready = 1;
+	}
+}
+
+static void hid_internal_hotplug_exit()
+{
+	if (!hid_hotplug_context.mutex_ready) {
+		return;
+	}
+
+	pthread_mutex_lock(&hid_hotplug_context.mutex);
+	hid_hotplug_callback** current = &hid_hotplug_context.hotplug_cbs
```

Comment:
I have no idea how did this wasn't fixed in the commit despite this code having actually been run and tested. Will look at what other changes didn't get commited.